external/storm-redis/src/main/java/org/apache/storm/redis/trident/state/RedisClusterMapState.java

```diff
@@ -0,0 +1,295 @@
+        public String hkey = null;
+    }
+
+    public static interface KeyFactory extends Serializable {
+        String build(List<Object> key);
+    }
+
+    /**
+     * OpaqueTransactional for redis-cluster.
+     * */
+    public static StateFactory opaque(JedisClusterConfig jedisClusterConfig) {
+        return opaque(jedisClusterConfig, new Options());
+    }
+
+    public static StateFactory opaque(JedisClusterConfig jedisClusterConfig, String hkey) {
+        Options opts = new Options();
+        opts.hkey = hkey;
+        return opaque(jedisClusterConfig, opts);
+    }
+
+    public static StateFactory opaque(JedisClusterConfig jedisClusterConfig, KeyFactory factory) {
+        Options opts = new Options();
+        opts.keyFactory = factory;
+        return opaque(jedisClusterConfig, opts);
+    }
+
+    public static StateFactory opaque(JedisClusterConfig jedisClusterConfig, Options<OpaqueValue> opts) {
+        return new Factory(jedisClusterConfig, StateType.OPAQUE, opts);
+    }
+
+    /**
+     * Transactional for redis-cluster.
+     * */
+    public static StateFactory transactional(JedisClusterConfig jedisClusterConfig) {
+        return transactional(jedisClusterConfig, new Options());
+    }
+
+    public static StateFactory transactional(JedisClusterConfig jedisClusterConfig, String hkey) {
+        Options opts = new Options();
+        opts.hkey = hkey;
+        return transactional(jedisClusterConfig, opts);
+    }
+
+    public static StateFactory transactional(JedisClusterConfig jedisClusterConfig, KeyFactory factory) {
+        Options opts = new Options();
+        opts.keyFactory = factory;
+        return transactional(jedisClusterConfig, opts);
+    }
+
+    public static StateFactory transactional(JedisClusterConfig jedisClusterConfig, Options<TransactionalValue> opts) {
+        return new Factory(jedisClusterConfig, StateType.TRANSACTIONAL, opts);
+    }
+
+    /**
+     * NonTransactional for redis-cluster.
+     * */
+    public static StateFactory nonTransactional(JedisClusterConfig jedisClusterConfig) {
+        return nonTransactional(jedisClusterConfig, new Options());
+    }
+
+    public static StateFactory nonTransactional(JedisClusterConfig jedisClusterConfig, String hkey) {
+        Options opts = new Options();
+        opts.hkey = hkey;
+        return nonTransactional(jedisClusterConfig, opts);
+    }
+
+    public static StateFactory nonTransactional(JedisClusterConfig jedisClusterConfig, KeyFactory factory) {
+        Options opts = new Options();
+        opts.keyFactory = factory;
+        return nonTransactional(jedisClusterConfig, opts);
+    }
+
+    public static StateFactory nonTransactional(JedisClusterConfig jedisClusterConfig, Options<Object> opts) {
+        return new Factory(jedisClusterConfig, StateType.NON_TRANSACTIONAL, opts);
+    }
+
+
+
+    protected static class Factory implements StateFactory {
+        // TODO : serialize redis.clients.jedis.JedisPoolConfig
```

Comment:
@revans2 
Actually it is encouraged to allow custom JedisPoolConfig or GenericObjectPoolConfig object for modifying Pool configuration, and then serialization is necessary.
But GenericObjectPoolConfig is from Apache Commons Pool 2, so it isn't easy way.

We can add some kind of setup methods from JedisPoolConfig and JedisClusterConfig which are exactly same to GenericObjectPoolConfig, to allow custom Pool setup, and generate GenericObjectPoolConfig or JedisPoolConfig object from these.

How do you think about?
cc. @dashengju